katib-controller crashed when parallelTrialCount is negative

/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
When I set experiment parallelTrialCount, the katib-contrller crashed with below message

```
/go/src/github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1337
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 285 [running]:
github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x154c500, 0x27fea20)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment.(*ReconcileExperiment).deleteTrials(0xc000709f90, 0xc000b8eb40, 0x2836ea8, 0x0, 0x0, 0x1, 0xc00087ec80, 0x11)
	/go/src/github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment/experiment_controller.go:340 +0x374
github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment.(*ReconcileExperiment).ReconcileTrials(0xc000709f90, 0xc000b8eb40, 0x2836ea8, 0x0, 0x0, 0xc0003a3c70, 0x0)
	/go/src/github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment/experiment_controller.go:264 +0x29b
github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment.(*ReconcileExperiment).ReconcileExperiment(0xc000709f90, 0xc000b8eb40, 0x2836ea8, 0x0)
	/go/src/github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment/experiment_controller.go:245 +0x37f
github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment.(*ReconcileExperiment).Reconcile(0xc000709f90, 0xc000a7d5f8, 0x8, 0xc000a7d640, 0xe, 0x2818f00, 0x0, 0x0, 0x0)
	/go/src/github.com/kubeflow/katib/pkg/controller.v1alpha3/experiment/experiment_controller.go:203 +0x518
github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00070a780, 0x0)
	/go/src/github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x1cc
github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/go/src/github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000b54b80)
	/go/src/github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000b54b80, 0x3b9aca00, 0x0, 0x1, 0xc000726660)
	/go/src/github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc000b54b80, 0x3b9aca00, 0xc000726660)
	/go/src/github.com/kubeflow/katib/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/src/github.com/kubeflow/katib/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x311

```
**What did you expect to happen:**


**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Kubeflow version:
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

--- a/pkg/controller.v1alpha3/experiment/util/status_util.go
+++ b/pkg/controller.v1alpha3/experiment/util/status_util.go
@@ -168,7 +168,7 @@ func UpdateExperimentStatusCondition(collector *ExperimentsCollector, instance *
 		return
 	}
 
-	if (instance.Spec.MaxFailedTrialCount != nil) && (failedTrialsCount >= *instance.Spec.MaxFailedTrialCount) {
+	if (instance.Spec.MaxFailedTrialCount != nil) && (failedTrialsCount > *instance.Spec.MaxFailedTrialCount) {
 		msg := "Experiment has failed because max failed count has reached"
 		instance.MarkExperimentStatusFailed(ExperimentFailedReason, msg)
 		instance.Status.CompletionTime = &now

--- a/pkg/webhook/v1alpha3/experiment/validation_webhook.go
+++ b/pkg/webhook/v1alpha3/experiment/validation_webhook.go
@@ -17,7 +17,9 @@ limitations under the License.
 package experiment
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"fmt"
 	"net/http"
 
@@ -51,9 +53,17 @@ func NewExperimentValidator(c client.Client) *experimentValidator {
 
 func (v *experimentValidator) Handle(ctx context.Context, req types.Request) types.Response {
 	inst := &experimentsv1alpha3.Experiment{}
+	var oldInst *experimentsv1alpha3.Experiment
 	err := v.decoder.Decode(req, inst)
 	if err != nil {
 		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+	if len(req.AdmissionRequest.OldObject.Raw) > 0 {
+		oldDecoder := json.NewDecoder(bytes.NewBuffer(req.AdmissionRequest.OldObject.Raw))
+		oldInst = &experimentsv1alpha3.Experiment{}
+		if err := oldDecoder.Decode(&oldInst); err != nil {
+			return admission.ErrorResponse(http.StatusBadRequest, fmt.Errorf("cannot decode incoming old object: %v", err))
+		}
 	}
 
 	// After metrics collector sidecar injection in Job level done, delete validation for namespace labels
@@ -75,7 +85,7 @@ func (v *experimentValidator) Handle(ctx context.Context, req types.Request) typ
 		return admission.ErrorResponse(http.StatusBadRequest, err)
 	}
 
-	err = v.ValidateExperiment(inst)
+	err = v.ValidateExperiment(inst, oldInst)
 	if err != nil {
 		return admission.ErrorResponse(http.StatusBadRequest, err)
 	}

--- a/pkg/webhook/v1alpha3/experiment/validator/validator.go
+++ b/pkg/webhook/v1alpha3/experiment/validator/validator.go
@@ -65,6 +65,16 @@ func (g *DefaultValidator) ValidateExperiment(instance *experimentsv1alpha3.Expe
 		return err
 	}
 
+	if instance.Spec.MaxFailedTrialCount != nil && *instance.Spec.MaxFailedTrialCount < 0 {
+		return fmt.Errorf("spec.maxFailedTrialCount should not be less than 0")
+	}
+	if instance.Spec.MaxTrialCount != nil && *instance.Spec.MaxTrialCount <= 0 {
+		return fmt.Errorf("spec.maxTrialCount must be greater than 0")
+	}
+	if instance.Spec.ParallelTrialCount != nil && *instance.Spec.ParallelTrialCount <= 0 {
+		return fmt.Errorf("spec.parallelTrialCount must be greater than 0")
+	}
+
 	return nil
 }
 

--- a/pkg/webhook/v1alpha3/experiment/validator/validator.go
+++ b/pkg/webhook/v1alpha3/experiment/validator/validator.go
@@ -8,6 +8,7 @@ import (
 	"strconv"
 	"strings"
 
+	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -23,7 +24,7 @@ import (
 var log = logf.Log.WithName("experiment-validating-webhook")
 
 type Validator interface {
-	ValidateExperiment(instance *experimentsv1alpha3.Experiment) error
+	ValidateExperiment(instance, oldInst *experimentsv1alpha3.Experiment) error
 	InjectClient(c client.Client)
 }
 
@@ -41,7 +42,26 @@ func (g *DefaultValidator) InjectClient(c client.Client) {
 	g.Generator.InjectClient(c)
 }
 
-func (g *DefaultValidator) ValidateExperiment(instance *experimentsv1alpha3.Experiment) error {
+func (g *DefaultValidator) ValidateExperiment(instance, oldInst *experimentsv1alpha3.Experiment) error {
+	if instance.Spec.MaxFailedTrialCount != nil && *instance.Spec.MaxFailedTrialCount < 0 {
+		return fmt.Errorf("spec.maxFailedTrialCount should not be less than 0")
+	}
+	if instance.Spec.MaxTrialCount != nil && *instance.Spec.MaxTrialCount <= 0 {
+		return fmt.Errorf("spec.maxTrialCount must be greater than 0")
+	}
+	if instance.Spec.ParallelTrialCount != nil && *instance.Spec.ParallelTrialCount <= 0 {
+		return fmt.Errorf("spec.parallelTrialCount must be greater than 0")
+	}
+	if oldInst != nil {
+		oldInst.Spec.MaxFailedTrialCount = instance.Spec.MaxFailedTrialCount
+		oldInst.Spec.MaxTrialCount = instance.Spec.MaxTrialCount
+		oldInst.Spec.ParallelTrialCount = instance.Spec.ParallelTrialCount
+		if equality.Semantic.DeepEqual(instance.Spec, oldInst.Spec) {
+			return nil
+		} else {
+			return fmt.Errorf("Only spec.parallelTrialCount, spec.maxTrialCount and spec.maxFailedTrialCount are editable.")
+		}
+	}
 	if err := g.validateObjective(instance.Spec.Objective); err != nil {
 		return err
 	}
@@ -64,17 +84,6 @@ func (g *DefaultValidator) ValidateExperiment(instance *experimentsv1alpha3.Expe
 	if err := g.validateMetricsCollector(instance); err != nil {
 		return err
 	}
-
-	if instance.Spec.MaxFailedTrialCount != nil && *instance.Spec.MaxFailedTrialCount < 0 {
-		return fmt.Errorf("spec.maxFailedTrialCount should not be less than 0")
-	}
-	if instance.Spec.MaxTrialCount != nil && *instance.Spec.MaxTrialCount <= 0 {
-		return fmt.Errorf("spec.maxTrialCount must be greater than 0")
-	}
-	if instance.Spec.ParallelTrialCount != nil && *instance.Spec.ParallelTrialCount <= 0 {
-		return fmt.Errorf("spec.parallelTrialCount must be greater than 0")
-	}
-
 	return nil
 }
 

--- a/pkg/webhook/v1alpha3/experiment/validator/validator_test.go
+++ b/pkg/webhook/v1alpha3/experiment/validator/validator_test.go
@@ -115,7 +115,7 @@ metadata:
 	}
 
 	for _, tc := range tcs {
-		err := g.ValidateExperiment(tc.Instance)
+		err := g.ValidateExperiment(tc.Instance, nil)
 		if !tc.Err && err != nil {
 			t.Errorf("Expected nil, got %v", err)
 		} else if tc.Err && err == nil {

